src/rgw/store/dbstore/common/dbstore.cc

```diff
@@ -1651,6 +1604,109 @@ int DB::Object::Write::write_data(const DoutPrefixProvider* dpp,
   return 0;
 }
 
+/*
+ * If versioned, 
+ * - check if the old version of the object already exists
+ *   if exists,
+ *     - set its version-id/instance to "null" if instance was empty
+ *     - demote it to NON_CURRENT 
+ * - create versioned object with FLAG_CURRENT set 
+ *
+ * XXX: call this function under a lock or sqlite transaction to prevent
+ * parallel object uploads marking both the versions as FLAG_CURRENT;
+ *
+ */
+int DB::Object::Write::write_versioned_obj(const DoutPrefixProvider *dpp, DBOpParams& params) {
+  DB *store = target->get_store();
+  int ret = -1;
+  DBOpParams obj_params = {};
+  uint64_t version_num = 0;
+
+  store->InitializeParams(dpp, &obj_params);
+  target->InitializeParamsfromObject(dpp, &obj_params);
+
+  /* Check if already exists */
+  obj_params.op.obj.state.obj.key.instance.clear();
+  ret = target->list_versioned_objects(dpp, obj_params.op.obj.list_entries);
+  
+  if (ret && ret != -ENOENT) {
+    ldpp_dout(dpp, 0)<<"ListVersionedObjects of object (" << obj_params.op.obj.state.obj.key.name << ") failed err:(" <<ret<<")" << dendl;
+    return ret;
+  }
+
+  for (auto& ent : obj_params.op.obj.list_entries) {
+    ret = target->update_obj_next_version(dpp, ent, false, version_num); // demote previous version
```

Comment:
yes. it is mainly used to add/clear FLAG_CURRENT flag & read the latest version-num used. This loop breaks as soon as it finds next version to demote. 
I thought this flag is necessary and is checked while processing any op (apart from listing and deleting the object if version-id not provided).  If thats not the case, then just for listing/deleting , we could list the object ordered by VersionNum and just read the top entry. 

This code should be protected by lock/sqlite-transaction which I haven't handled yet. If this flag is not used any 

> is there any way that `write_versioned_obj()` could avoid this loop to update all other versions? if this is just to clear the `FLAG_CURRENT`, can we find a way to make the 'current' version an implicit property, instead of tracking it explicitly in this flag?

yes. it is mainly used to add/clear FLAG_CURRENT flag & read the last version-num used. This loop breaks as soon as it finds next version to demote/promote. 
I thought this flag is necessary and is checked while processing any op (apart from listing and deleting the object if version-id not provided).  If that's not the case, then just for listing/deleting , we could list the object ordered by VersionNum and just read the top entry, maybe .

Reply:
i *believe* the only time this CURRENT flag matters is for [ListObjectVersions](https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListObjectVersions.html), because it has to return a boolean `IsLatest` for each entry. it also comes up with lifecycle for current vs. noncurrent expiration, but we implement that on top of this bucket listing so have that same `IsLatest` info available

given a listing with `ORDER BY ObjName, VersionNum DESC`, we could assign an implicit `IsLatest` based on whether its the first entry with the given `ObjName`. the only place this doesn't work is when the `ListObjectVersions` request includes a `version-id-marker` to resume a listing in the middle of a given `ObjName`'s versions. in this case, we might need to issue a separate query like `SELECT VersionNum FROM {object-table} WHERE ObjName = {key-marker} ORDER BY VersionNum DESC LIMIT 1` to determine whether this `VersionNum` matches the first entry returned by the other listing (essentially what you said: "we could list the object ordered by VersionNum and just read the top entry")

Reply:
ok thanks! 
seems like AUTOINCREMENT can be used only if there is one INT primary key. So instead I am checking if we can make use of sqlite triggers to update version number atomically post adding object entry.

Reply:
'sqlite trigger' helped to increment version_num atomically. Updated the PR with the changes.
I also now handled verison_supended case. Kindly review.

https://github.com/ceph/s3-tests/pull/458/files contains the tests list which now pass with this versioning support in dbstore.

Reply:
@dang @cbodley @mattbenjamin .. kindly review the latest changes!